[Gutenberg] Add "Featured Destinations" homepage block

miina opened the pull request:
~~Based on #12 and #31.~~

Aims to do the following:
- [x] Pull data dynamically to editor.
- [x] Sorts and renders the terms according to the uploaded images.
- [x] Register and add relevant meta to `location` term.

miina committed:
Register featured block. Pull data dynamically. Merge #31

miina commented:
@DavidCramer Started adding the featured block within this commit: 9f745ae
Basically just registered the block and meta for `amp_travel_featured` and enabled fetching the terms via API by the value of the meta. Also added the logic for when there are not enough featured locations found.
 
What's missing:
- Rendering for when there are enough featured locations found (Completely missing for frontend in PHP and partly missing in JS for editor);
- Adding image-related metafield(s) for the user to upload the image;
- `Location` term field to mark it being `Featured`;
- Sorting the images for putting together the grid.

miina committed:
Fix API request URL.

DavidCramer committed:
Fix typo

miina committed:
Add img meta to locations. Sort REST response for the editor.

miina committed:
Add JS file for term controls.

miina committed:
Merge branch 'feature/25-featured_block' of github.com:xwp/travel into feature/25-featured_block

miina committed:
Merge develop & resolve conflicts.

miina committed:
Add featured destination meta.

miina committed:
Merge develop & fix conflicts.

miina committed:
Add front-end PHP render for featured block.

miina committed:
Fix editor-side rendering.

kienstra commented on includes/class-amp-travel-taxonomies.php:
```diff
@@ -22,6 +29,108 @@ public function init() {
+		<div class='form-field form-required location-cover-image-wrap'>
+			<label for='location-cover-image'><?php esc_html_e( 'Activity Cover Image' ); ?></label>
+			<input type='hidden' id='location-cover-image-value' class='small-text' name='location-cover-image-value' value='' />
+			<input type='button' id='location-cover-image' class='button location-cover-image-upload-button' value='<?php esc_html_e( 'Upload' ); ?>' />
```
Could the "Upload" text for the button instead be something like "Select"? It's also possible to select an image from the media library.

<img width="458" alt="upload-image-button" src="https://user-images.githubusercontent.com/4063887/38756479-46f7d908-3f2f-11e8-924b-ddf6778ffd90.png">

kienstra replied:
Thanks, the new "Select" text looks good.

kienstra commented on assets/js/term-controls.js:
```diff
@@ -0,0 +1,82 @@
+
+		} );
+
+		component.removeButton.on( 'click', function( event ) {
```
If you have time, could you display this `component.removeButton` only if there's no selected image? But I'd call this a "nice to have."

It could be confusing to have a "Remove" button if there's no image to remove.

kienstra replied:
Thanks for making this change, @miina. This works well.

kienstra commented on assets/js/term-controls.js:
```diff
@@ -0,0 +1,82 @@
+	/**
+	 * Init.
+	 */
+	component.init = function init() {
```
It looks like the selected image is persisting even after saving a new "Location"

1. Go to `/wp-admin/edit-tags.php?taxonomy=location`
2. Fill in all of the fields, and select an image
<img width="1069" alt="adding-new-location" src="https://user-images.githubusercontent.com/4063887/38754681-6f08e538-3f28-11e8-9a09-9ade2aaacaf7.png">

3. Click "Add New Location"
4. Expected: the selected image doesn't display anymore, as the rest of the fields are empty
5. Actual: the selected image still displays:
<img width="1115" alt="image-still-appears" src="https://user-images.githubusercontent.com/4063887/38754755-aa893694-3f28-11e8-89ef-93842059b898.png">

But the workflow looks fine when editing a location:

<img width="1120" alt="edit-location" src="https://user-images.githubusercontent.com/4063887/38754843-fe6bbdfe-3f28-11e8-952d-d37d3fea051e.png">

kienstra replied:
Thanks for fixing this.

kienstra commented on blocks/featured/index.js:
```diff
@@ -93,9 +23,95 @@ export default registerBlockType(
+				featuredLocations: '/wp/v2/locations?per_page=6&meta_value=1&meta_key=amp_travel_featured'
+			};
+		} )( ( { featuredLocations } ) => { // eslint-disable-line
+			const hasLocations = Array.isArray( featuredLocations.data ) && 6 === featuredLocations.data.length;
```
What do you think about making this 
```javascript
featuredLocations.data.length >= 6
```
instead of:
```javascript
6 === featuredLocations.data.length
```

It might be easy for someone to change `AMP_Travel_Blocks::$featured_locations_count`, without noticing that this line requires it to be exactly `6`.

kienstra commented on includes/class-amp-travel-taxonomies.php:
```diff
@@ -22,6 +29,108 @@ public function init() {
+	 */
+	public function add_location_meta_fields() {
+		?>
+		<div class='form-field form-required location-cover-image-wrap'>
```
Could this method please use double quotes for the attributes, like:
```php
<div class="form-field form-required location-cover-image-wrap">
```

And likewise for `edit_location_meta_fields()`

kienstra commented on includes/class-amp-travel-taxonomies.php:
```diff
@@ -22,6 +29,108 @@ public function init() {
+	 * @param integer $term_id Term ID.
+	 */
+	public function save_location_meta( $term_id ) {
+
```
Could you please remove this empty line?

kienstra commented on includes/class-amp-travel-taxonomies.php:
```diff
@@ -22,6 +29,108 @@ public function init() {
+	 * @param WP_Term $term WP_Term being edited.
+	 */
+	public function edit_location_meta_fields( $term ) {
+		$cover_img_id = get_term_meta( $term->term_id, 'amp_travel_location_img', true );
```
Could you please store `amp_travel_location_img` in a constant, like `AMP_Travel_Taxonomies::LOCATION_IMG`. Then, this class and `AMP_Travel_Blocks` could access it.

kienstra commented on includes/class-amp-travel-taxonomies.php:
```diff
@@ -22,6 +29,108 @@ public function init() {
+	 */
+	public function save_location_meta( $term_id ) {
+
+		if ( ! wp_verify_nonce( $_POST['travel_location_meta_nonce'], basename( __FILE__ ) ) ) {
```
The nonce in `$_POST['travel_location_meta_nonce']` doesn't look to be set when saving a new location at `/wp-admin/edit-tags.php?taxonomy=location`. And the image and "Featured destination" selection don't look to be saved.

But maybe there's a different result in your environment.

kienstra replied:
Thanks, `save_location_meta()` looks to be working.

kienstra commented on assets/js/term-controls.js:
```diff
@@ -0,0 +1,82 @@
+( function( $ ) {
+	'use strict';
+
+	let component = {};
```
Nice structure of the file.

kienstra commented on assets/js/term-controls.js:
```diff
@@ -0,0 +1,82 @@
+
+			// Create the media frame.
+			component.fileFrame[ id ] = wp.media.frames.fileFrame = wp.media( {
+				title: self.data( 'uploader_title' ),
```
`self.data( 'uploader_title' )` and ` self.data( 'uploader_button_text' )` are `undefined` when debugging on the page `/wp-admin/edit-tags.php?taxonomy=location`.

kienstra commented on includes/class-amp-travel-blocks.php:
```diff
@@ -403,4 +404,274 @@ public function render_block_discover( $attributes ) {
+			'taxonomy'   => 'location',
+			'meta_key'   => 'amp_travel_featured',
+			'meta_value' => 1,
+			'per_page'   => self::$featured_locations_count,
```
`'per_page' => self::$featured_locations_count`
 should probably be
`'number' => self::$featured_locations_count`

kienstra commented on includes/class-amp-travel-blocks.php:
```diff
@@ -403,4 +404,274 @@ public function render_block_discover( $attributes ) {
+		foreach ( $terms as $term_array ) {
+
+			// If the input might be as objects as well.
+			if ( ! is_array( $term_array ) ) {
```
How about changing `$term_array` to `$term`, as it might not be an array?

kienstra commented on blocks/featured/index.js:
```diff
@@ -93,9 +23,95 @@ export default registerBlockType(
 			__( 'Travel' )
 		],
 
-		edit() {
```
This [is outside](https://github.com/xwp/travel/pull/34/files#diff-f13e039fde62e426638518f5cefb1783L87) of this PR's diff. But could the title of this block be 'Featured Destinations,' instead of 'Featured'? Or something that's descriptive.
```javascript
title: __( 'Featured Destinations' ),
category: 'common',
icon: 'wordpress-alt',
```

kienstra replied:
Thanks for making this change.

kienstra commented on includes/class-amp-travel-taxonomies.php:
```diff
@@ -22,6 +29,108 @@ public function init() {
+			<td>
+				<?php wp_nonce_field( basename( __FILE__ ), 'travel_location_meta_nonce' ); ?>
+
+				<input type='hidden' id='location-cover-image-value' class='small-text' name='location-cover-image-value' value='<?php esc_attr( $cover_img_id ); ?>' />
```
```php
<?php esc_attr( $cover_img_id ); ?>
```
This could probably use an `echo` at the beginning:
```php
<?php echo esc_attr( $cover_img_id ); ?>
```

kienstra commented on includes/class-amp-travel-blocks.php:
```diff
@@ -403,4 +404,274 @@ public function render_block_discover( $attributes ) {
+			$location_img_srcset = wp_get_attachment_image_srcset( $location_img_id, 'full' );
+
+			$output .= '<a href="' . esc_url( get_term_link( $location['term_id'] ) ) . '" class="travel-featured-tile flex flex-auto relative travel-featured-color-' .
+							esc_html( $location_params[ $i ]['color'] ) . '">
```
```php
esc_html( $location_params[ $i ]['color'] 
```
Could this use `esc_attr()` instead, as this is for an attribute (class)

kienstra commented on includes/class-amp-travel-blocks.php:
```diff
@@ -403,4 +404,274 @@ public function render_block_discover( $attributes ) {
+							esc_html( $location_params[ $i ]['color'] ) . '">
+							<amp-img class="travel-object-cover flex-auto" layout="responsive" width="' .
+							esc_html( $location_params[ $i ]['width'] ) . '" height="' .
+							esc_html( $location_params[ $i ]['height'] ) . '" srcset="' . esc_html( $location_img_srcset ) . '" src="' . esc_url( $location_img_src[0] ) . '""></amp-img>
```
Could this line and the line above use `esc_attr()` for all of the attributes? But maybe there was an issue in using `esc_attr()` with the `srcset`.

kienstra commented on blocks/featured/index.js:
```diff
@@ -93,9 +23,95 @@ export default registerBlockType(
+								icon="admin-post"
+								label={ __( 'Locations' ) }
+					>
+						{ __( 'Not enough featured locations found, add at least six to use the block.' ) }
```
Would it be easier to use if the message mentioned "Location" terms, maybe like:
```
Not enough featured locations found. Please add at least six "Locations" terms, select an image, and check "Featured destination."
```

kienstra commented on includes/class-amp-travel-blocks.php:
```diff
@@ -12,6 +12,13 @@
+	 *
+	 * @var int
+	 */
+	public static $featured_locations_count = 6;
```
This class is instantiated, so would there be a need to make this static? If not, could you please remove `static`, and change this to something like:
```php
const FEATURED_LOCATIONS_COUNT = 6;
```

Then, references to it could change from `self::$featured_locations_count` to `self::FEATURED_LOCATIONS_COUNT`

This is outside the diff of this PR, but could you please do the same thing to `$popular_posts_count` [below](https://github.com/xwp/travel/pull/34/files#diff-d85bb5c137cc4b60d13bd558d64a58d9R27)?

kienstra reviewed (changes requested):
**Nice Work**
*Some Points Below*

Hi @miina,
Good work here. Especially in setting up the grid:

<img width="1253" alt="featured-destinations-grid" src="https://user-images.githubusercontent.com/4063887/38761933-982baa94-3f4c-11e8-9ca0-4f7faa7f7513.png">

kienstra commented:
**Editor Styling**

This might be outside the scope of this PR. But we might revisit the styling of this block in the editor. The images appear much taller than on the front-end.

miina committed:
Fixes from review. Improve editor style.

miina committed:
Add additional destinations number check.

miina commented:
@kienstra Thanks for all the found issues! These should be addressed now, also improved the Gutenberg styling a little bit.

kienstra commented:
**Approved, Other Than Merge Conflicts**

Hi @miina,
Thanks for all of your work here. And thanks improving the styling of this "Featured Destinations" block in the editor. 

I caused these merge conflicts by merging in #16. Could you please resolve them? 

I tried to do that on my own, but it looks like [editor-blocks.js](https://github.com/xwp/travel/blob/develop/assets/js/editor-blocks.js) will need to be rebuilt. And this command didn't seem to rebuild it properly:

```
npm run build
```

I probably missed something 😄 

Thanks!

miina committed:
Merge develop & resolve conflicts.

miina commented:
@kienstra, Merged develop & resolved conflicts, sorry for not noticing the conflicts before.

miina committed:
Add compiled editor JS

kienstra committed:
Change featured/index.js save() to return null.

Before, there was an error in Gutenberg,
as renderStaticFeaturedBlock() wasn't defined.
This PR removed that function.
So change save() to return null,
as this looks to be a dynamic block.
AMP_Travel_Blocks::render_block_travel_featured()
renders the output on the front-end.
Also, compile js/editor-blocks.js, using:
npm run build.

kienstra commented:
**Merging, Request For Sanity Check**

Hi @miina,
Thanks, this block looks great. No need to apologize for not noticing the merge conflicts. I introduced them after your work day.

Could you please sanity-check [this commit](https://github.com/xwp/travel/pull/34/commits/b8a9d4133e29481541e709aad62a81a79fecc31e)? I changed the `save()` function to return `null`. And I recompiled `editor-blocks.js` with `npm run build`.

miina commented:
Hey @kienstra, the commit you did looks good, thanks for that!
